### ISSUE TITLE
Reference stealing

### ISSUE THREAD

**Issue by jakirkham** [cython/cython#2722]
Opening this issue to discuss how Cython handles reference stealing. Have noted that some functions account for this by returning `PyObject*` instead of `object`. However with arguments this strategy [doesn't seem to work correctly]( https://stackoverflow.com/q/41925500 ). Namely the `PyObject*` will still result in the `object`'s reference count decremented and may require an explicit cast. Given this, am raising this issue to figure out how we should handle cases where the reference is stolen. This comes up with some Python and NumPy API functions that have this behavior (e.g. in-place operations and macros).

**Comment by scoder** [comment]
Good question. :) Let me collect a couple of thoughts on this.

First of all, regarding the examples at hand, the obvious solution for users is to not use `PyList_SetItem(lst, i, obj)` but `lst[i] = obj`. For a typed list, Cython can optimise the latter but not the first (so the assignment is likely faster than the C-API call), and the code readability also suffers a lot from such misoptimisations.

Then, it feels to me like this is not really a bug in Cython, because Cython does the right thing, seen from the point of the caller. It's the function itself that misbehaves, in a way. It is still a usability issue, because Cython can't help with this kind of contract and does not guard against accidental misuse.

Changing the argument type would be a signal to the user that something unusual is going on, and that it's up to the user to take care of it. It's not perfect, because `PyObject*` is just a clumsy way to say `object`, and users could just replace `func(obj)` with `func(<PyObject*>obj)` to make it compile again, but not a single bit safer. It would raise the bar, at least. But it's also something that will be difficult to get away from again if we later decide that we want Cython to be more helpful.

But then, what could Cython do? Automatically increase the reference count of the argument when it gets passed? That would go against the intention of these functions, which specifically want to avoid useless reference counting.

What Cython could do is to set the reference to `NULL` after the call (without decrefing it) and then generating NULL checking code wherever necessary when the variable is used. Note that the call might happen inside of a loop, if-block, try-except/finally and whatnot, and the variable might also be part of a closure, so further usage might happen in various places. We have a mechanism for this, it's used for closure variables, since they may or may not have been initialised in other places that use it. But it also increases the runtime overhead, meaning that calling a function on a variable could introduce totally unpredictable slow downs in other parts of the code. Not great at all.

Personally, I think we'd be better off entirely without such functions. They are just too error prone by design.

I'd be happy to hear other thoughts on this, but could live with just raising the bar for users of such functions by requiring an explicit cast to make them aware that they are on their own.

**Comment by robertwb** [comment]
I agree that all reference-stealing function should be marked with a large warning. 

The problem is that there's not really a "correct" way to indicate this with argument types as the way to "properly" use these functions is all about what you do (or don't do) after passing the argument. This makes it hard (impossible?) to check via types. 

It's worth noting that there is a certain convenience here for writing against the raw C API, as one can pass new references to these functions without worrying about cleaning them up. Increfing before the call might not be that bad (and try to optimize it away if it matters).

**Comment by scoder** [comment]
I added more visible warnings to the functions in 9d2ae85b2b5ae36deb68132152a56437e5cf9615, although I doubt that many people will read those, if they didn't even see the big grey boxes in the CPython docs.

I think, in the end, users who directly call C-API functions are on their own and should better be knowing what they are doing. (Which doesn't mean that we can't find a way to deal with this general issue in Cython in some way, so I'll leave this ticket open.)

**Comment by CaselIT** [comment]
Not sure if continuing here is the best approach. If not I can open a new issue.

Just encountered this behavior on 3.0.5 when using `PyList_New`.

I think at the very least the fact that cython does not automatically increases the reference when calling functions that steal a reference could be documented (searching for `steal` or `PyList_SetItem` in the website yields no result, not sure if it's under other keywords)
 
> First of all, regarding the examples at hand, the obvious solution for users is to not use PyList_SetItem(lst, i, obj) but lst[i] = obj. For a typed list, Cython can optimise the latter but not the first (so the assignment is likely faster than the C-API call), and the code readability also suffers a lot from such misoptimisations.

Using `lst[i] = obj` seems to not be an option in this case, since it segfault either way. In the example below `call_with_setitem` works correctly once Py_INCREF is added, while `call_with_assign` segfault either way

```cython
from cpython.list cimport PyList_New, PyList_SetItem
from cpython.ref cimport Py_INCREF

cdef list with_setitem(int num):
    cdef list res = PyList_New(num)
    cdef Py_ssize_t i
    for i in range(num):
        y = int(i*i)
        Py_INCREF(y)
        PyList_SetItem(res, i, y)
    return res

def call_with_setitem(value):
    return with_setitem(value)

cdef list with_assign(int num):
    cdef list res = PyList_New(num)
    cdef Py_ssize_t i
    for i in range(num):
        y = int(i*i)
        Py_INCREF(y)
        res[i] = y # segfault with or without Py_INCREF
    return res

def call_with_assign(value):
    return with_assign(value)
```

**Comment by jakirkham** [comment]
Recognize this may be an illustrative example for a more complex use case. That said, it can be worth considering generator expressions with `list`s where Cython often generates the performant code that is desired while allowing for arguably better readability for future devs. Of course this may or may not be possible depending on the code in practice

```cython
cdef list with_gen(Py_ssize_t n):
    cdef Py_ssize_t i
    cdef list res = [i*i for i in range(n)]
    return res
```

Side note: Typed the function's input argument to match the looping variable

**Comment by scoder** [comment]
> Using `lst[i] = obj` seems to not be an option in this case, since it segfault either way.

That is because you are creating a list in an invalid state using `PyList_New()`. It returns a list with an empty array (all `NULL` values), so you cannot fill that list with normal Python operations. If you decide to use `PyList_New()` yourself, you have to go all the way and fill the list using corresponding C-API functions. And for that, you have to read their documentation.

I second the suggestion that a list comprehension is probably fast enough in most cases. If you really find that it isn't, then Cython allows you to use the C-API instead, but you have to take care yourself to use it correctly.

As I wrote in my first comment above, it's really not easy for Cython to guess your intentions here, and guessing always makes the behaviour less predictable.

**Comment by da-woods** [comment]
It'd possibly be useful to provide something like `cython.steal` so you could do

```
cdef list with_setitem(int num):
    cdef list res = PyList_New(num)
    cdef Py_ssize_t i
    for i in range(num):
        y = int(i*i)
        PyList_SetItem(res, i, cython.steal(y))
    return res
```

It'd effectively just do `del y` but without the decref (i.e. set the pointer to 0 and let Cython's flow control know that it no longer exists), which would let people that want to write manual code use a bit less reference counting.

**Comment by CaselIT** [comment]
> As I wrote in my first comment above, it's really not easy for Cython to guess your intentions here, and guessing always makes the behaviour less predictable.

I completely agree, but I believe that the documentation does a poor job of indicating this.
It may be obvious for experienced users but for less experienced ones is not clear and searching the documentation does not offer any indication regarding this
